prediction_market_agent/agents/microchain_agent/nft_treasury_game/scripts/generate_report.py

```diff
@@ -0,0 +1,97 @@
+import textwrap
+
+import typer
+from langchain_core.prompts import PromptTemplate
+from prediction_market_agent_tooling.gtypes import ChecksumAddress, xdai_type
+from prediction_market_agent_tooling.loggers import logger
+from prediction_market_agent_tooling.tools.balances import get_balances
+from prediction_market_agent_tooling.tools.contract import (
+    ContractOwnableERC721OnGnosisChain,
+)
+from prediction_market_agent_tooling.tools.parallelism import par_map
+from web3 import Web3
+
+from prediction_market_agent.agents.microchain_agent.memory import DatedChatMessage
+from prediction_market_agent.agents.microchain_agent.memory_functions import (
+    fetch_memories_from_last_run,
+)
+from prediction_market_agent.agents.microchain_agent.nft_treasury_game.constants_nft_treasury_game import (
+    NFT_TOKEN_FACTORY,
+)
+from prediction_market_agent.agents.microchain_agent.nft_treasury_game.deploy_nft_treasury_game import (
+    DEPLOYED_NFT_AGENTS,
+)
+from prediction_market_agent.agents.utils import _summarize_learnings
+from prediction_market_agent.db.models import LongTermMemories
+
+SUMMARY_PROMPT_TEMPLATE = """
+Summarize the memories below. They contain summarizations of the actions of AI agents competing on an NFT game.
+
+Memories:
+{memories}"""
+
+FINAL_SUMMARY_PROMPT_TEMPLATE = """
+Make a final summary of a collection of summaries from each agent. Encompass the main activities that took place involving the agents.
+
+Memories:
+{memories}"""
+
+
+def get_nft_balance(owner_address: ChecksumAddress, web3: Web3) -> int:
+    contract = ContractOwnableERC721OnGnosisChain(
+        address=Web3.to_checksum_address(NFT_TOKEN_FACTORY)
+    )
+    balance: int = contract.balanceOf(
+        Web3.to_checksum_address(owner_address), web3=web3
+    )
+    return balance
+
+
+def summarize_past_actions_from_agent(agent_memories: list[LongTermMemories]) -> str:
+    dated_chat_messages = [
+        DatedChatMessage.from_long_term_memory(ltm) for ltm in agent_memories
+    ]
+    prompt = PromptTemplate.from_template(SUMMARY_PROMPT_TEMPLATE)
+    learnings_per_agent = _summarize_learnings(
+        memories=[str(m) for m in dated_chat_messages], prompt_template=prompt
+    )
+    return learnings_per_agent
+
+
+def summarize_prompts_from_all_agents() -> str:
+    memories_last_run = fetch_memories_from_last_run(
+        agent_identifiers=[i.identifier for i in DEPLOYED_NFT_AGENTS]
+    )
+    # We generate the learnings from each agent's memories.
+    learnings: list[str] = par_map(
+        items=list(memories_last_run.values()), func=summarize_past_actions_from_agent
+    )
```

Comment:
We generate a per-agent summary because `stuff` method of load_summarize_chain (https://python.langchain.com/api_reference/langchain/chains/langchain.chains.summarize.chain.load_summarize_chain.html#load-summarize-chain), used by `summarize_learnings`, cannot process more than 100k tokens (gpt-4o context window).
I tried a bit with map-reduce and refine (see doc above), but would have required tons of refactoring. Doing like this is less complex and gets the result we are after, i.e. a summarization of the game run.